bbl_screen-patch/kexec_ui/printerui/qml/dialog/KexecDialog.qml

```diff
@@ -16,34 +18,81 @@ Item {
+        }
+    }())
+     
+    property var startupStrings: {
+        "default": { /* Normal Boot: 10 sec dialog then boot X1Plus */
+            "buttons": {
+                "yes_confirm": {"text": qsTr("Boot X1Plus"), "action": function() { X1PlusNative.system("/opt/kexec/boot"); }},
+                "no": {"text": qsTr("Startup options..."), "action": function() { dialogStack.replace("../BootOptionsPage.qml", {hasSdCard: hasSdCard}) }},
+                "cancel": {"text": "", "action": function() {}},  
+            },
+            "title": qsTr("Bootable SD card detected."),
+            "subtitle": function() {
+                return qsTr("Your printer will automatically boot from the SD card in %1 seconds.").arg(countdown)
+            }
+        },
+        "ota": {
+            "buttons": {/* auto install x1p Boot: 10 sec dialog then tell InstallingPage.qml to start installing x1p */
+                "yes_confirm": {"text": qsTr("Install"), "action": function() { dialogStack.replace("../InstallingPage.qml", {x1pName: x1pName}) }},
+                "no": {"text": qsTr("Startup options"), "action": function() { dialogStack.replace("../BootOptionsPage.qml", {hasSdCard: hasSdCard}) }},
+                "cancel": {"text": qsTr("Boot X1Plus"), "action": function() { X1PlusNative.system("/opt/kexec/boot"); }}
+            },
+            "title": qsTr("Update ready to install!"),
+            "subtitle": function() {
+                return qsTr("Automatically installing .x1p file '%1' in %2 seconds. To cancel or skip installation, select 'Startup options' or 'Boot X1Plus'.").arg(x1pName).arg(countdown);
+            }
+        }
+    }
 
     property var buttons: SimpleItemModel {
         DialogButtonItem {
-            name: "yes_confirm"; title: qsTr("Start from SD card")
+            name: "yes_confirm"
+            title: startupStrings[startupMode].buttons["yes_confirm"].text
             isDefault: defaultButton == 0
-            onClicked: { countdown = 0 }
+            onClicked: startupStrings[startupMode].buttons["yes_confirm"].action()
             visible: hasSdCard && countdown > 0
         }
         DialogButtonItem {
-            name: "no"; title: qsTr("Startup options...")
+            name: "no"
+            title: startupStrings[startupMode].buttons["no"].text
             isDefault: defaultButton == 1
-            onClicked: { timer.stop(); dialogStack.replace("../BootOptionsPage.qml", {hasSdCard: hasSdCard}); }
+            onClicked: startupStrings[startupMode].buttons["no"].action()
             visible: countdown > 0
         }
+        DialogButtonItem {
+            name: "cancel"
+            title: startupStrings[startupMode].buttons["cancel"].text
+            isDefault: defaultButton == 2
+            onClicked: startupStrings[startupMode].buttons["cancel"].action()
+            visible: hasSdCard && countdown > 0 && startupStrings[startupMode].buttons["cancel"].text.length > 0
+        }
+    }
```

Comment:
I feel like all this logic is trying to be three (?) different dialogs when it is in fact not three different dialogs...?

Reply:
> I feel like all this logic is trying to be three (?) different dialogs when it is in fact not three different dialogs...?

There are just two sets of dialog. No dialog to display if the first stage installer is running. This is overly complicated though, you are right about that.